src/theme.c

```diff
@@ -101,9 +136,28 @@ create_hover_fallback(struct theme *theme, struct lab_data_buffer **hover_buffer
 	cairo_paint(cairo);
 
 	/* Overlay (non-multiplied alpha) */
-	set_cairo_color(cairo, (float[4]) { 0.5f, 0.5f, 0.5f, 0.3f});
-	cairo_rectangle(cairo, 0, 0, width, height);
-	cairo_fill(cairo);
+	float overlay_color[4] = { 0.5f, 0.5f, 0.5f, 0.3f};
+	enum corner corner = corner_from_icon_name(icon_name);
+
+	if (corner == LAB_CORNER_UNKNOWN) {
+		set_cairo_color(cairo, overlay_color);
+		cairo_rectangle(cairo, 0, 0, width, height);
+		cairo_fill(cairo);
+	} else {
+		struct rounded_corner_ctx rounded_ctx = {
+			.box = &(struct wlr_box) { .width = width, .height = height },
+			.radius = rc.corner_radius,
+			.line_width = theme->border_width,
+			.fill_color = overlay_color,
+			.border_color = overlay_color,
+			.corner = corner
+		};
+		struct lab_data_buffer *overlay_buffer = rounded_rect(&rounded_ctx);
```

Comment:
In `create_hover_fallback()`, could we just use `rounded_rect()` for all cases? Just change .radius and .corner depending on `LAB_CORNER_UNKNOWN/TOP_LEFT/TOP_RIGHT`.

I didn't study it for long, but my eyes were drawn to the creation of a `struct lab_data_buffer` for only purpose of calling `cairo_set_source_surface()` and `cairo_paint()`.

Reply:
> In `create_hover_fallback()`, could we just use `rounded_rect()` for all cases? Just change .radius and .corner depending on `LAB_CORNER_UNKNOWN/TOP_LEFT/TOP_RIGHT`.

Theoretically yes. However I didn't see the need for it (which includes creating a new cairo / lab_data_buffer buffer) if we have a cairo primitive which does the same.

> I didn't study it for long, but my eyes were drawn to the creation of a `struct lab_data_buffer` for only purpose of calling `cairo_set_source_surface()` and `cairo_paint()`.

Well, `rounded_rect()` returns a `lab_data_buffer` which we then paint over the centered icon on transparent background. We could change `rounded_rect()` to return a cairo buffer directly but I don't think that would really ease code maintenance nor would it improve performance. We could however supply an existing cairo buffer to `rounded_rect()` and if `!= NULL` the `lab_data_buffer` creation is skipped and the function draws directly onto the given cairo surface and returns `NULL`.

Reply:
Good points. Let’s leave as is.